File: apps/mobile/lib/features/progress/domain/entities/progress.dart
/// Entités du domaine progression (immuables, écrites à la main).
library;

/// Période d'analyse des statistiques.
enum ProgressPeriod {
  week('week', 'Semaine'),
  month('month', 'Mois'),
  year('year', 'Année');

  const ProgressPeriod(this.apiValue, this.label);

  final String apiValue;
  final String label;
}

/// Point de graphique : volume soulevé sur un intervalle.
class ProgressPoint {
  const ProgressPoint({
    required this.bucketStart,
    required this.sessionsCount,
    required this.volumeKg,
  });

  final DateTime bucketStart;
  final int sessionsCount;
  final double volumeKg;
}

/// Statistiques agrégées d'une période.
class ProgressOverviewEntity {
  const ProgressOverviewEntity({
    required this.period,
    required this.sessionsCount,
    required this.setsCount,
    required this.totalVolumeKg,
    required this.totalDurationSeconds,
    required this.points,
  });

  final ProgressPeriod period;
  final int sessionsCount;
  final int setsCount;
  final double totalVolumeKg;
  final int totalDurationSeconds;
  final List<ProgressPoint> points;
}

enum PersonalRecordType {
  maxWeight('MAX_WEIGHT', 'Charge max'),
  maxReps('MAX_REPS', 'Répétitions max'),
  maxSetVolume('MAX_SET_VOLUME', 'Volume max sur une série');

  const PersonalRecordType(this.apiValue, this.label);

  final String apiValue;
  final String label;

  static PersonalRecordType fromApi(String value) =>
      PersonalRecordType.values.firstWhere(
        (type) => type.apiValue == value,
        orElse: () => PersonalRecordType.maxWeight,
      );
}

/// Record personnel sur un exercice.
class PersonalRecordEntry {
  const PersonalRecordEntry({
    required this.id,
    required this.exerciseName,
    required this.type,
    required this.value,
    required this.achievedAt,
    this.exerciseId,
    this.reps,
    this.weightKg,
  });

  final String id;
  final String? exerciseId;
  final String exerciseName;
  final PersonalRecordType type;

  /// kg, répétitions ou kg de volume selon [type].
  final double value;
  final int? reps;
  final double? weightKg;
  final DateTime achievedAt;

  String get formattedValue {
    final rounded = value == value.roundToDouble()
        ? value.round().toString()
        : value.toStringAsFixed(1);
    return type == PersonalRecordType.maxReps ? '$rounded rép.' : '$rounded kg';
  }
}

enum BodyMetricKind {
  weightKg('WEIGHT_KG', 'Poids (kg)'),
  bodyFatPercent('BODY_FAT_PERCENT', 'Masse grasse (%)');

  const BodyMetricKind(this.apiValue, this.label);

  final String apiValue;
  final String label;

  static BodyMetricKind fromApi(String value) =>
      BodyMetricKind.values.firstWhere(
        (kind) => kind.apiValue == value,
        orElse: () => BodyMetricKind.weightKg,
      );
}

/// Mesure corporelle datée (poids, masse grasse…).
class BodyMetricEntry {
  const BodyMetricEntry({
    required this.id,
    required this.kind,
    required this.value,
    required this.measuredAt,
  });

  final String id;
  final BodyMetricKind kind;
  final double value;
  final DateTime measuredAt;
}

/// Un point de la progression sur UN exercice : une séance, sa meilleure
/// charge et son volume.
///
/// Le serveur sert `maxWeightKg` et `maxReps` à `null` quand la séance n'a
/// porté aucune charge (poids du corps, cardio) : l'absence se dit, elle ne
/// se remplace pas par un zéro qui ressemblerait à un échec.
class ExerciseProgressionPoint {
  const ExerciseProgressionPoint({
    required this.sessionId,
    required this.date,
    required this.volumeKg,
    this.maxWeightKg,
    this.maxReps,
    this.distanceMeters = 0,
    this.durationSeconds = 0,
  });

  final String sessionId;
  final DateTime date;
  final double volumeKg;
  final double? maxWeightKg;
  final int? maxReps;

  /// Ce que la séance a parcouru et chronométré sur cet exercice, SOMMÉ :
  /// trois fractionnés de 400 m font 1 200 m de course. Zéro sur un
  /// exercice de fonte, ce qui est exact.
  final int distanceMeters;
  final int durationSeconds;

  /// Vrai quand la séance a laissé une trace de cardio.
  bool get hasCardio => distanceMeters > 0 || durationSeconds > 0;
}

/// La progression sur un exercice : ses séances ET ses records, ensemble.
///
/// Les deux arrivent dans la même réponse parce qu'ils se lisent ensemble :
/// un record est un point de la courbe, pas une liste à côté.
class ExerciseProgressionEntity {
  const ExerciseProgressionEntity({
    required this.exerciseId,
    required this.exerciseName,
    required this.records,
    required this.points,
  });

  final String exerciseId;
  final String exerciseName;
  final List<PersonalRecordEntry> records;

  /// Du plus ancien au plus récent, prêts pour un tracé.
  final List<ExerciseProgressionPoint> points;

  /// Séances où une charge a été notée : les seules traçables en kilos.
  List<ExerciseProgressionPoint> get chargedPoints =>
      points.where((point) => point.maxWeightKg != null).toList();

  /// Séances où une distance ou un chrono a été noté.
  List<ExerciseProgressionPoint> get cardioPoints =>
      points.where((point) => point.hasCardio).toList();

  /// Ce que cet exercice raconte le mieux : des kilos, ou des kilomètres.
  ///
  /// Le choix se fait sur les FAITS, pas sur une étiquette d'exercice : une
  /// fiche mal catégorisée n'a alors aucune conséquence, et un exercice
  /// hybride (le rameur chargé, la marche lestée) suit ce qu'on y a
  /// réellement noté. À égalité, la charge l'emporte — c'est le cœur de
  /// l'application.
  bool get readsAsCardio => cardioPoints.length > chargedPoints.length;

  /// Ce qu'une courbe cardio trace : la distance quand elle est notée au
  /// moins aussi souvent que le chrono, le temps sinon.
  ///
  /// La distance l'emporte à égalité parce qu'elle dit la performance : un
  /// coureur qui met le même temps sur plus de kilomètres progresse, et
  /// l'inverse ne se lit pas.
  CardioReading get cardioReading {
    final avecDistance = cardioPoints
        .where((point) => point.distanceMeters > 0)
        .length;
    final avecChrono = cardioPoints
        .where((point) => point.durationSeconds > 0)
        .length;
    return avecDistance >= avecChrono
        ? CardioReading.distance
        : CardioReading.duration;
  }
}

/// Ce qu'une courbe cardio met en ordonnée.
enum CardioReading { distance, duration }

File: packages/api-contracts/src/progress.ts
import { z } from 'zod';

/** Contrats de la progression (/api/v1/progress, /api/v1/body-metrics). */

export const progressPeriodSchema = z.enum(['week', 'month', 'year']);
export type ProgressPeriod = z.infer<typeof progressPeriodSchema>;

/** Point de série pour les graphiques (regroupement par jour/semaine/mois). */
export const progressPointSchema = z.object({
  bucketStart: z.string(),
  sessionsCount: z.number(),
  volumeKg: z.number(),
});
export type ProgressPoint = z.infer<typeof progressPointSchema>;

export const progressOverviewSchema = z.object({
  period: progressPeriodSchema,
  from: z.string(),
  to: z.string(),
  sessionsCount: z.number(),
  setsCount: z.number(),
  totalVolumeKg: z.number(),
  totalDurationSeconds: z.number(),
  points: z.array(progressPointSchema),
});
export type ProgressOverview = z.infer<typeof progressOverviewSchema>;

/**
 * Ce que la VIE ENTIÈRE compte, pour les récompenses.
 *
 * Des FAITS, jamais une règle : le serveur dit combien de séances et
 * quelles semaines, le moteur de récompenses (mobile) décide seul ce qu'est
 * une « meilleure série » ou une « semaine équilibrée ». Calculer ces règles
 * ici les dupliquerait, et deux copies divergent.
 *
 * Pourquoi le serveur : le mobile les dérivait de son historique LOCAL,
 * plafonné à 60 séances au rapatriement. Sur un compte à 200 séances, un
 * téléphone neuf en voyait 60 — et une récompense déjà gagnée disparaissait.
 */
export const lifetimeWeekSchema = z.object({
  /** Lundi qui ouvre la semaine, `YYYY-MM-DD` dans le fuseau de la personne. */
  mondayOn: z.string(),
  /** Séances TERMINÉES cette semaine-là. */
  sessions: z.number(),
});
export type LifetimeWeek = z.infer<typeof lifetimeWeekSchema>;

export const lifetimeStatsSchema = z.object({
  completedSessions: z.number(),
  /** Une entrée par semaine ACTIVE, de la plus ancienne à la plus récente. */
  weeks: z.array(lifetimeWeekSchema),
});
export type LifetimeStats = z.infer<typeof lifetimeStatsSchema>;

export const personalRecordTypeSchema = z.enum(['MAX_WEIGHT', 'MAX_REPS', 'MAX_SET_VOLUME']);
export type PersonalRecordType = z.infer<typeof personalRecordTypeSchema>;

export const personalRecordSchema = z.object({
  id: z.string(),
  exerciseId: z.string().nullable(),
  exerciseName: z.string(),
  recordType: personalRecordTypeSchema,
  /** Valeur du record : kg, répétitions ou kg de volume selon le type. */
  value: z.number(),
  reps: z.number().nullable(),
  weightKg: z.number().nullable(),
  achievedAt: z.string(),
});
export type PersonalRecord = z.infer<typeof personalRecordSchema>;

/** Progression sur un exercice : meilleure charge et volume par séance. */
export const exerciseProgressionPointSchema = z.object({
  sessionId: z.string(),
  date: z.string(),
  maxWeightKg: z.number().nullable(),
  maxReps: z.number().nullable(),
  volumeKg: z.number(),
  /**
   * Ce que la séance a parcouru et chronométré sur CET exercice.
   *
   * SOMMÉS, jamais maximisés : trois fractionnés de 400 m font 1 200 m de
   * course, alors qu'une charge ne s'additionne pas d'une série à l'autre.
   * Zéro sur un exercice de fonte, ce qui est exact — un tapis et un développé
   * couché ne se lisent pas sur la même courbe, et c'est le client qui choisit
   * laquelle tracer.
   */
  distanceMeters: z.number(),
  durationSeconds: z.number(),
});
export type ExerciseProgressionPoint = z.infer<typeof exerciseProgressionPointSchema>;

export const exerciseProgressionSchema = z.object({
  exerciseId: z.string(),
  exerciseName: z.string(),
  records: z.array(personalRecordSchema),
  points: z.array(exerciseProgressionPointSchema),
});
export type ExerciseProgression = z.infer<typeof exerciseProgressionSchema>;

export const bodyMetricTypeSchema = z.enum(['WEIGHT_KG', 'BODY_FAT_PERCENT']);
export type BodyMetricType = z.infer<typeof bodyMetricTypeSchema>;

export const bodyMetricSchema = z.object({
  id: z.string(),
  metricType: bodyMetricTypeSchema,
  value: z.number(),
  measuredAt: z.string(),
});
export type BodyMetric = z.infer<typeof bodyMetricSchema>;

/**
 * Correction d'une mesure déjà enregistrée : la valeur, la date, ou les deux.
 *
 * Le TYPE ne se corrige pas — un poids ne devient pas un taux de masse
 * grasse. Se tromper de type se répare en supprimant la ligne et en en
 * créant une autre, ce que l'API permet déjà.
 *
 * `refine` plutôt que deux champs obligatoires : un corps vide n'est pas une
 * correction, c'est un appel qui ne veut rien dire. Le refuser tôt évite une
 * écriture inutile et un `updatedAt` qui bougerait pour rien.
 */
export const updateBodyMetricRequestSchema = z
  .object({
    value: z.number().min(1).max(500).optional(),
    measuredAt: z.string().optional(),
  })
  .refine(
    (body) => body.value !== undefined || body.measuredAt !== undefined,
    'Rien à corriger : donne au moins la valeur ou la date.',
  );
export type UpdateBodyMetricRequest = z.infer<typeof updateBodyMetricRequestSchema>;
